amethyst_assets/src/simple.rs

```diff
@@ -4,9 +4,80 @@
 use std::borrow::Cow;
 use std::error::Error;
 use std::marker::PhantomData;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use parking_lot::RwLock;
 
 use {Asset, AssetSpec, Cache, Context};
 
+/// An `AssetPtr` which provides `push_update`, `update`
+/// and `is_shared` methods. These can simply be called
+/// in order to implement the `Asset` trait.
+///
+/// The recommended strategy is to create a struct for an
+/// asset which simply wraps `AssetPtr` and implements
+/// `Asset` by calling these methods. Methods for the asset
+/// can then be implemented on that wrapper struct by getting
+/// the inner asset with `inner` and `inner_mut`.
+#[derive(Clone)]
+pub struct AssetPtr<A> {
+    inner: A,
+    update: Arc<(AtomicUsize, RwLock<Option<A>>)>,
+    version: usize,
+}
+
+impl<A> AssetPtr<A> {
+    /// Creates a new asset pointer.
+    pub fn new(data: A) -> Self {
+        AssetPtr {
+            inner: data,
+            update: Arc::new((AtomicUsize::new(0), RwLock::new(None))),
+            version: 0,
+        }
+    }
+
+    /// Borrows the inner asset.
+    pub fn inner(&self) -> &A {
+        &self.inner
+    }
+
+    /// Borrows the inner asset mutably.
+    pub fn inner_mut(&mut self) -> &mut A {
+        &mut self.inner
+    }
+}
+
+impl<A> AssetPtr<A>
+    where A: Clone
+{
+    /// Pushes an update to the shared update container;
+    /// this update can then be applied to all asset pointers by calling
+    /// `update` on them.
+    pub fn push_update(&self, updated: Self) {
+        let &(ref count, ref lock) = &*self.update;
+
+        *lock.write() = Some(updated.inner);
+        count.fetch_add(1, Ordering::SeqCst);
```

Comment:
@torkleyy If you make the changes to `Release` and `Acquire` then I consider this ready to merge.